backends/memcache.go

```diff
@@ -67,44 +69,108 @@ func (m *Memcache) Execute(cmd string, args ...interface{}) (interface{}, error)
 	return "inside GEt", nil
 }
 
-func (m *Memcache) Delete(keys ...interface{}) int {
-	return 1
+func (m *Memcache) Delete(keys ...interface{}) (int, error) {
+	mc := m.GetClient()
+	for _, key := range keys {
+		skey := key.(string)
+		err := mc.Delete(skey)
+		if err != nil {
+			return 0, err
+		}
+	}
+	m.PutClient(mc)
+	return 1, nil
 }
 
-func (m *Memcache) Get(key string) string {
+func (m *Memcache) Get(key string) (string, error) {
 	mc := m.GetClient()
 	it, erm := mc.Get(key)
 	m.PutClient(mc)
 	if erm != nil {
-		errMsg := fmt.Sprintf("Error in getting key %s", key)
-		return errMsg
+		return "", erm
 	}
-	return string(it.Value)
+	return string(it.Value), nil
 }
 
-func (m *Memcache) Set(key string, value interface{}) bool {
+func (m *Memcache) Set(key string, value interface{}) (bool, error) {
+	svalue := value.(string)
 	mc := m.GetClient()
-	newVal := value.(string)
-	erm := mc.Set(&memcache.Item{Key: key, Value: []byte(newVal)})
+	erm := mc.Set(&memcache.Item{Key: key, Value: []byte(svalue)})
 	m.PutClient(mc)
 	if erm != nil {
-		return false
+		fmt.Println(erm)
+		return false, erm
+	}
+	return true, nil
+}
+
+func (m *Memcache) MSet(keyValMap map[string]interface{}) (bool, error) {
+	return false, nil
+}
+
+func (m *Memcache) MGet(keys ...interface{}) ([]string, error) {
+	skeys := make([]string, len(keys))
+	for _, key := range keys {
+		skeys = append(skeys, key.(string))
+	}
+	mc := m.GetClient()
+	items, err := mc.GetMulti(skeys)
+	m.PutClient(mc)
+	if err != nil {
+		return []string{}, nil
 	}
-	return true
+	arr := make([]string, 10)
+	for _, key := range skeys {
+		if item, ok := items[key]; ok {
+			arr = append(arr, string(item.Value))
+		} else {
+			arr = append(arr, "")
+		}
+	}
+	return arr, nil
+}
+
+func (m *Memcache) Expire(key string, duration int) (bool, error) {
+	mc := m.GetClient()
+	err := mc.Touch(key, int32(duration))
+	m.PutClient(mc)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
 }
 
-func (m *Memcache) MGet(keys ...interface{}) []string {
-	return []string{"hello world"}
+func (m *Memcache) Setex(key string, duration int, val interface{}) (bool, error) {
+	mc := m.GetClient()
+	defer m.PutClient(mc)
+	sval := val.(string)
+	erm := mc.Set(&memcache.Item{Key: key, Value: []byte(sval)})
+	if erm != nil {
+		return false, erm
+	} else {
+		err := mc.Touch(key, int32(duration))
+
+		if err != nil {
+			return false, err
+		}
+	}
+	return true, nil
 }
 
-func (m *Memcache) MSet(mapOfKeyVal map[string]interface{}) bool {
-	return true
+
+// All Set methods raise unimplemented error.
+// Raise unimplemented error
+func (m *Memcache) Smembers(key string) (set.Set, error) {
```

Comment:
I am not sure if we should be using external "set" library here. Now when clients call this method, they have to use this "set" library to work with the output.

Can we use slice?. Set related operations is anyway handled by Redis, so whatever smembers returns, we are sure they are going to unique elements. Doesn't make much sense to use a "set library" in language as well.